src/modules/bulletproofs/rangeproof_impl.h

```diff
@@ -0,0 +1,792 @@
+        secp256k1_scalar_set_int(&one, 1);
+        secp256k1_lr_generate(&lr_gen, &l, &r, &one);
+        secp256k1_scalar_mul(&l, &l, &r);
+        secp256k1_scalar_add(&t1, &t1, &l);
+    }
+
+    /* B = t0 - t1 + t2 = l(-1) dot r(-1) */
+    secp256k1_lr_generator_init(&lr_gen, nonce, &y, &z, nbits, value, min_value, n_commits);
+    secp256k1_scalar_clear(&t2);
+    for (i = 0; i < nbits * n_commits; i++) {
+        secp256k1_scalar negone;
+        secp256k1_scalar l, r;
+        secp256k1_scalar_set_int(&negone, 1);
+        secp256k1_scalar_negate(&negone, &negone);
+        secp256k1_lr_generate(&lr_gen, &l, &r, &negone);
+        secp256k1_scalar_mul(&l, &l, &r);
+        secp256k1_scalar_add(&t2, &t2, &l);
+    }
+
+    /* t1 = (A - B)/2 */
+    secp256k1_scalar_set_int(&tmps, 2);
+    secp256k1_scalar_inverse_var(&tmps, &tmps);
+    secp256k1_scalar_negate(&t2, &t2);
+    secp256k1_scalar_add(&t1, &t1, &t2);
+    secp256k1_scalar_mul(&t1, &t1, &tmps);
+
+    /* t2 = -(-B + t0) + t1 */
+    secp256k1_scalar_add(&t2, &t2, &t0);
+    secp256k1_scalar_negate(&t2, &t2);
+    secp256k1_scalar_add(&t2, &t2, &t1);
+
+    /* Compute Ti = t_i*A + tau_i*G for i = 1,2 */
+    /* TODO surely we can improve this */
+    secp256k1_ecmult_const(&tmpj, value_gen, &t1, 256);
+    secp256k1_ge_set_gej(&out_pt[2], &tmpj);
+    secp256k1_ecmult_const(&tmpj, &gens->blinding_gen[0], &tau1, 256);
+    secp256k1_gej_add_ge(&tmpj, &tmpj, &out_pt[2]);
+    secp256k1_ge_set_gej(&out_pt[2], &tmpj);
+
+    secp256k1_ecmult_const(&tmpj, value_gen, &t2, 256);
+    secp256k1_ge_set_gej(&out_pt[3], &tmpj);
+    secp256k1_ecmult_const(&tmpj, &gens->blinding_gen[0], &tau2, 256);
+    secp256k1_gej_add_ge(&tmpj, &tmpj, &out_pt[3]);
+    secp256k1_ge_set_gej(&out_pt[3], &tmpj);
+
+    /* get challenge x */
+    secp256k1_bulletproof_update_commit(commit, &out_pt[2], &out_pt[3]);
+    secp256k1_scalar_set_b32(&x, commit, &overflow);
+    if (overflow || secp256k1_scalar_is_zero(&x)) {
+        return 0;
+    }
+    secp256k1_scalar_sqr(&xsq, &x);
+
+    /* compute tau_x and mu */
+    secp256k1_scalar_mul(&taux, &tau1, &x);
+    secp256k1_scalar_mul(&tmps, &tau2, &xsq);
+    secp256k1_scalar_add(&taux, &taux, &tmps);
+    for (i = 0; i < n_commits; i++) {
+        secp256k1_scalar_mul(&tmps, &zsq, &blind[i]);
+        secp256k1_scalar_add(&taux, &taux, &tmps);
+        secp256k1_scalar_mul(&zsq, &zsq, &z);
+    }
+
+    secp256k1_scalar_mul(&mu, &rho, &x);
+    secp256k1_scalar_add(&mu, &mu, &alpha);
+
+    /* Negate taux and mu so the verifier doesn't have to */
+    secp256k1_scalar_negate(&taux, &taux);
+    secp256k1_scalar_negate(&mu, &mu);
+
+    /* Encode rangeproof stuff */
+    secp256k1_scalar_get_b32(&proof[0], &taux);
+    secp256k1_scalar_get_b32(&proof[32], &mu);
+    secp256k1_bulletproof_serialize_points(&proof[64], out_pt, 4);
+
+    /* Mix this into the hash so the input to the inner product proof is fixed */
+    /* TODO is this necessary? revisit */
```

Comment:
Hm, good question. It seems to be a good idea to me because it ensures that an inner product proof cannot be copied and reused in a different context. Maybe @bbuenz has more insights... The question is if it's necessary to hash the first part of the range proof into the challenge for the inner product proof.

Reply:
Yes I think so. Otherwise the prover can choose the first IP challenge before he knows P. The way I have always done it is by hashing the last challenge into the next challenge.